### ISSUE TITLE
REDUCED

### ISSUE THREAD

**Issue by jindrichmynarz** [w3c/sparql-12#67]
## Why?

The [`REDUCED`](https://www.w3.org/TR/sparql11-query/#modReduced) solution modifier is underspecified in SPARQL 1.1. Current SPARQL engines treat it *"like a 'best effort' DISTINCT"* ([source](https://stackoverflow.com/questions/2990343/sparql-distinct-vs-reduced/2990586#2990586)). What constitutes the best effort differs between the implementations, which makes queries using `REDUCED` unportable. The query `SELECT REDUCED ?o WHERE { VALUES ?o { 0 1 1 0} }` results in `0, 1, 1, 0` in Virtuoso, `0, 1, 0` in Fuseki and GraphDB, and `0, 1` in Blazegraph. Indeed, under-defined features cannot be implemented consistently, except by chance or via an out-of-spec consensus. Consequently, [SPARQL 1.1 Test Suite](https://www.w3.org/2009/sparql/docs/tests/README.html) asks for a special lax handing of tests using `REDUCED` allowing duplicate solutions.

Perhaps due to its variable interpretation and limited portability queries using `REDUCED` are rare. Since its specification doesn't give users useful guarantees, they don't know what to expect from it and be reluctant to use it. Analyzing [Wikidata query logs](https://analytics.wikimedia.org/datasets/one-off/wikidata/sparql_query_logs/), `REDUCED` appears in 0.0004 % of 654M queries. [Occurences of `REDUCED`](https://github.com/search?q=%22SELECT+REDUCED%22+language%3ASPARQL&type=Code) in code hosted by Github often appear in tests rather than in business logic.

## Previous work

Some query engines, such as Fuseki and GraphDB, already treat `REDUCED` as removing consecutive duplicates.

## Proposed solution

`REDUCED` can be either soft-deprecated, or specified as removing consecutive duplicates. A no-op resolution can document the variability of `REDUCED` and clarify that consumers are responsible for deduplication of query results.

## Considerations for backward compatibility

Specification of underspecified behaviour breaks only the consumers that rely on specific implementation behaviour.

**Comment by rubensworks** [comment]
> REDUCED can be either soft-deprecated, or specified as removing consecutive duplicates.

Another solution may be to add a parameter to `REDUCED` that defines the window in which duplicates should be found, e.g.:

* `REDUCED(n)`: Removes duplicates within a sliding window of the `n` last bindings.

`REDUCED` _without_ param could then be left for implementors to choose a certain default param.
To remain compatible with the mentioned results above, Virtuoso could pick `0` as default, Fuseki `1` and GraphDB `2` (or more).

_(This is how `REDUCED` is implemented in Comunica, with that difference that the sliding window size `n` should be defined at configuration-time, instead of at query-time)_

**Comment by kal** [comment]
I agree that the under-specification of REDUCED is an issue and limits the practical applicability of the keyword. The solutions above would at least better specify what REDUCED actually means, but they would still not guarantee cross-engine compatibility as they are dependent on the (engine-specific) order in which bindings are generated.

**Comment by RickMoynihan** [comment]
I agree REDUCED is underspecified; however I consider it more a query hint; that it is permissable but need not be guaranteed to remove solutions if they are known to be duplicated.  

However I actually ran into a similar issue with stardog some time ago where it did no deduping via reduced at all; permissable perhaps, but no use.  I filed it as a bug; and it was quickly implemented by deduping adjacent results.  This fixed a real production issue for me, where a modest amount of real results was being duplicated through cartesian joins... iirc it was an orders of magnitude saving something like n!.

So I certainly don't support soft deprecating this feature, as it's useful.  However tightening the spec, or at least saying it `SHOULD` as a minimum dedupe adjacent solutions (which is as kal says still be engine dependent, but it's at least specifiable) would be something I'd support; though I see no need to rule out other strategies for reducing results further; without requiring implementers to consume the whole result set.  Importantly with `REDUCED` it is possible to stream results.

**Comment by TallTed** [comment]
Picking a nit --
> the (engine-specific) order in which bindings are generated

The order in which bindings are generated is not necessarily (though it might be) engine-specific.  Lacking any `ORDER BY` clause, the order of bindings must be treated as entirely indeterminate, even with the same data being served by the same engine.
